### PR TITLE
Remove kafka lookup records when a record is tombstoned

### PR DESCRIPTION
**Pull request by hwball**
### Description

Lookups made using the `druid-kafka-extraction-namespace` extension do not currently handle records that have been deleted from a kafka topic. Deleting a record from a kafka topic involves sending a record with a null payload and the same key as the record you want to delete to the topic. This is also known as sending a tombstone. This PR aims to handle key value records from kafka that have keys but also null values.

On receiving a key value record that has a key and a null value, the corresponding record stored in the factory map will be removed. 

<hr>

##### Key changed/added classes in this PR
 * `KafkaLookupExtractorFactory.java`

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
Travis didn't trigger for some reason. I closed and re-opened the PR. changes LGTM. I will merge once CI passes.

**Review by paul-rogers — COMMENTED** [review]
Great fix, thanks!

Should we document this behavior in the docs? The only reference I could find is [here](https://druid.apache.org/docs/0.23.0/querying/lookups.html#api-for-configuring-lookups), but that doesn't address streaming updates.

**Comment by cloventt** [comment]
The docs are actually buried in the [extensions area](docs/development/extensions-core/kafka-extraction-namespace.md). I've updated them to mention this new behaviour.

File: extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java
Comment (abhishekagarwal87) [inline extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java:284]: why not use `10` here as well?

Comment (abhishekagarwal87) [reply extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java:284]: @hwball @cloventt - can you address this comment? Last thing to do before we can merge this PR and https://github.com/apache/druid/pull/12842

Comment (cloventt) [reply extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java]: Thanks @abhishekagarwal87 I've pushed that change, sorry for the delay.

**Comment by abhishekagarwal87** [comment]
thank you. I will merge this once CI is green.

File: extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java
Comment (hwball) [reply extensions-core/kafka-extraction-namespace/src/test/java/org/apache/druid/query/lookup/TestKafkaExtractionCluster.java]: Looks like it fails on CI with a lower sleep time. It was tested locally a few times and passed with the lower value but I will revert the change so the CI can pass.

**Comment by abhishekagarwal87** [comment]
@hwball @cloventt  - Thank you for your contribution. I have merged the PR.

**Pull request merged**
